### PR TITLE
Add tool to analyze pod runtimes from jobs

### PR DESCRIPTION
**Pull request by alvaroaleman**
Output looks like this:
```
$ /tmp/runtime-analzer
All runtimes
+-----------------------------------+-----------------------+---------+
|                POD                |       CONTAINER       | RUNTIME |
+-----------------------------------+-----------------------+---------+
| private-org-peribolos-sync-build  | manage-dockerfile     | 0s      |
| private-org-sync-build            | manage-dockerfile     | 0s      |
| determinize-peribolos-build       | manage-dockerfile     | 0s      |
| ci-operator-prowgen-build         | manage-dockerfile     | 0s      |
| ci-operator-config-mirror-build   | manage-dockerfile     | 0s      |
| src-build                         | manage-dockerfile     | 0s      |
| determinize-prow-jobs-build       | manage-dockerfile     | 0s      |
| auto-peribolos-sync-build         | manage-dockerfile     | 0s      |
| determinize-ci-operator-build     | manage-dockerfile     | 0s      |
| secret-wrapper-build              | manage-dockerfile     | 0s      |
| autobumper-build                  | manage-dockerfile     | 0s      |
| dptp-controller-manager-build     | manage-dockerfile     | 0s      |
| config-shard-validator-build      | manage-dockerfile     | 0s      |
| autoowners-build                  | manage-dockerfile     | 0s      |
| pj-rehearse-build                 | manage-dockerfile     | 0s      |
| determinize-prow-config-build     | manage-dockerfile     | 0s      |
| group-auto-updater-build          | manage-dockerfile     | 0s      |
| result-aggregator-build           | manage-dockerfile     | 0s      |
| ci-operator-build                 | manage-dockerfile     | 0s      |
| config-change-trigger-build       | manage-dockerfile     | 0s      |
| config-brancher-build             | manage-dockerfile     | 0s      |
| registry-replacer-build           | manage-dockerfile     | 0s      |
| ci-operator-checkconfig-build     | manage-dockerfile     | 0s      |
| bugzilla-backporter-build         | manage-dockerfile     | 0s      |
| ipi-deprovision-build             | manage-dockerfile     | 0s      |
| repo-brancher-build               | manage-dockerfile     | 0s      |
| blocking-issue-creator-build      | manage-dockerfile     | 0s      |
| ci-secret-bootstrap-build         | manage-dockerfile     | 0s      |
| ci-operator-configresolver-build  | manage-dockerfile     | 0s      |
| bin-build                         | manage-dockerfile     | 1s      |
| format                            | test                  | 1s      |
| private-prow-configs-mirror-build | manage-dockerfile     | 1s      |
| auto-config-brancher-build        | manage-dockerfile     | 1s      |
| testgrid-config-generator-build   | manage-dockerfile     | 1s      |
| applyconfig-build                 | manage-dockerfile     | 1s      |
| repo-init-build                   | manage-dockerfile     | 1s      |
| sanitize-prow-jobs-build          | manage-dockerfile     | 1s      |
| src-build                         | extract-image-content | 7s      |
| integration                       | test                  | 11s     |
| secret-wrapper-build              | docker-build          | 19s     |
| validate-vendor                   | test                  | 20s     |
| private-org-peribolos-sync-build  | docker-build          | 33s     |
| repo-brancher-build               | docker-build          | 33s     |
| format                            | artifacts             | 35s     |
| private-prow-configs-mirror-build | docker-build          | 40s     |
| determinize-ci-operator-build     | docker-build          | 40s     |
| result-aggregator-build           | docker-build          | 41s     |
| config-brancher-build             | docker-build          | 42s     |
| blocking-issue-creator-build      | docker-build          | 43s     |
| group-auto-updater-build          | docker-build          | 45s     |
| sanitize-prow-jobs-build          | docker-build          | 45s     |
| bugzilla-backporter-build         | docker-build          | 45s     |
| testgrid-config-generator-build   | docker-build          | 49s     |
| autobumper-build                  | docker-build          | 50s     |
| private-org-sync-build            | docker-build          | 50s     |
| pj-rehearse-build                 | docker-build          | 51s     |
| repo-init-build                   | docker-build          | 55s     |
| ci-operator-build                 | docker-build          | 1m0s    |
| integration                       | artifacts             | 1m0s    |
| ci-operator-prowgen-build         | docker-build          | 1m4s    |
| unit                              | test                  | 1m6s    |
| ci-secret-bootstrap-build         | docker-build          | 1m6s    |
| determinize-peribolos-build       | docker-build          | 1m14s   |
| determinize-prow-config-build     | docker-build          | 1m18s   |
| applyconfig-build                 | docker-build          | 1m19s   |
| config-shard-validator-build      | docker-build          | 1m21s   |
| determinize-prow-jobs-build       | docker-build          | 1m21s   |
| ci-operator-config-mirror-build   | docker-build          | 1m21s   |
| ci-operator-checkconfig-build     | docker-build          | 1m23s   |
| breaking-changes                  | test                  | 1m25s   |
| dptp-controller-manager-build     | docker-build          | 1m26s   |
| auto-peribolos-sync-build         | docker-build          | 1m29s   |
| ci-operator-configresolver-build  | docker-build          | 1m44s   |
| src-build                         | docker-build          | 1m49s   |
| unit                              | artifacts             | 1m50s   |
| validate-vendor                   | artifacts             | 1m50s   |
| config-change-trigger-build       | docker-build          | 1m54s   |
| breaking-changes                  | artifacts             | 1m55s   |
| autobumper-build                  | extract-image-content | 2m1s    |
| secret-wrapper-build              | extract-image-content | 2m1s    |
| repo-brancher-build               | extract-image-content | 2m7s    |
| pj-rehearse-build                 | extract-image-content | 2m7s    |
| ci-operator-build                 | extract-image-content | 2m7s    |
| ci-operator-configresolver-build  | extract-image-content | 2m7s    |
| ci-secret-bootstrap-build         | extract-image-content | 2m7s    |
| private-org-sync-build            | extract-image-content | 2m10s   |
| autoowners-build                  | docker-build          | 2m11s   |
| registry-replacer-build           | docker-build          | 2m16s   |
| ipi-deprovision-build             | extract-image-content | 2m20s   |
| auto-config-brancher-build        | docker-build          | 2m23s   |
| private-org-peribolos-sync-build  | extract-image-content | 2m24s   |
| ipi-deprovision-build             | docker-build          | 3m26s   |
| ci-operator-config-mirror-build   | extract-image-content | 3m50s   |
| testgrid-config-generator-build   | extract-image-content | 3m51s   |
| ci-operator-checkconfig-build     | extract-image-content | 3m51s   |
| config-shard-validator-build      | extract-image-content | 3m52s   |
| determinize-prow-jobs-build       | extract-image-content | 3m53s   |
| determinize-peribolos-build       | extract-image-content | 3m55s   |
| config-change-trigger-build       | extract-image-content | 3m55s   |
| auto-peribolos-sync-build         | extract-image-content | 3m55s   |
| auto-config-brancher-build        | extract-image-content | 4m0s    |
| determinize-prow-config-build     | extract-image-content | 4m1s    |
| repo-init-build                   | extract-image-content | 4m1s    |
| autoowners-build                  | extract-image-content | 4m2s    |
| dptp-controller-manager-build     | extract-image-content | 4m12s   |
| applyconfig-build                 | extract-image-content | 4m20s   |
| config-brancher-build             | extract-image-content | 5m23s   |
| group-auto-updater-build          | extract-image-content | 5m24s   |
| bugzilla-backporter-build         | extract-image-content | 5m25s   |
| result-aggregator-build           | extract-image-content | 5m26s   |
| blocking-issue-creator-build      | extract-image-content | 5m28s   |
| registry-replacer-build           | extract-image-content | 5m28s   |
| ci-operator-prowgen-build         | extract-image-content | 5m29s   |
| determinize-ci-operator-build     | extract-image-content | 5m30s   |
| sanitize-prow-jobs-build          | extract-image-content | 5m31s   |
| private-prow-configs-mirror-build | extract-image-content | 5m31s   |
| bin-build                         | docker-build          | 6m28s   |
+-----------------------------------+-----------------------+---------+
Runtimes by container
+-----------------------+--------+-----+-------+-----------------+------------+
|       CONTAINER       | MEDIAN | MIN |  MAX  |      STDEV      | SAMPLESIZE |
+-----------------------+--------+-----+-------+-----------------+------------+
| manage-dockerfile     | 0s     | 0s  | 1s    | 395.772412ms    |         36 |
| test                  | 20s    | 1s  | 1m25s | 32.879172738s   |          5 |
| artifacts             | 1m50s  | 35s | 1m55s | 32.465366161s   |          5 |
| extract-image-content | 3m55s  | 7s  | 5m31s | 1m24.274546475s |         35 |
| docker-build          | 1m5s   | 19s | 6m28s | 1m4.017141801s  |         36 |
+-----------------------+--------+-----+-------+-----------------+------------+
```

Not completely sure if this is useful right now

### CODE REVIEW

**Comment by petr-muller** [comment]
IMO the interesting bit is the overhead of `extract-image-content`. The actual image builds are quite fast but this piece of setup is slow.

File: pkg/job-runtime-analyzer/job-runtime-analyzer.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	for _, item := range data {
+		table.Append([]string{item.podName, item.containerName, item.duration.String()})
+	}
+	for _, footer := range footers {
+		table.SetFooter(footer)
+	}
+	table.Render()
+}
+
+// award for worst named type?
```
Comment (petr-muller) [inline pkg/job-runtime-analyzer/job-runtime-analyzer.go:97]: Statistics calls these tuples a "summary" :)

**Comment by petr-muller** [comment]
Also, can we get a readme?

**Comment by alvaroaleman** [comment]
@petr-muller yeah so at this point I was mostly asking myself how and if to continue here. As you correctly noted whats most interesting is the `extract-image-content` container. In order for this to be actually useful we would IMHO need to get the data for many job runs to see if it changed over time and also to be able to compare 4.X was 3.X. Then we are suddenly at the point where we don't know what question we have exactly anymore and probably just want a json dump with a summary that we can then re-use within something else that is better suited at exploring statistical data (Python, as much as I hate to say it ).

File: pkg/job-runtime-analyzer/job-runtime-analyzer.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+	podList := &corev1.PodList{}
+	if err := json.Unmarshal(rawPodList, podList); err != nil {
+		return fmt.Errorf("failed to unmarshal pod list: %w", err)
+	}
+
+	var runtimes []*podContainerRuntime
+	for _, pod := range podList.Items {
+		for _, container := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
+			if pod.Name == "test" {
```
Comment (droslean) [inline pkg/job-runtime-analyzer/job-runtime-analyzer.go:53]: This check shouldn't be in this loop.

File: pkg/job-runtime-analyzer/job-runtime-analyzer.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	printRuntimes("All runtimes", runtimes)
+	printRuntimeByContainer(runtimesByContainer)
+
+	return nil
+}
+
+func printRuntimes(title string, data []*podContainerRuntime, footers ...[]string) {
+	_, _ = fmt.Printf("%s\n", title)
```
Comment (droslean) [inline pkg/job-runtime-analyzer/job-runtime-analyzer.go:85]: is this a debugging leftover?

Comment (alvaroaleman) [reply pkg/job-runtime-analyzer/job-runtime-analyzer.go:85]: No

File: pkg/job-runtime-analyzer/job-runtime-analyzer.go (hunk 2 above)
Comment (alvaroaleman) [reply pkg/job-runtime-analyzer/job-runtime-analyzer.go:53]: The check is just wrong, should be `container.Name`

File: cmd/job-runtime-analyzer/main.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"flag"
+
+	"github.com/sirupsen/logrus"
+
+	jobruntimeanalyzer "github.com/openshift/ci-tools/pkg/job-runtime-analyzer"
+)
+
+const defaultPodJSONURL = "https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_ci-tools/978/pull-ci-openshift-ci-tools-master-images/1280941429257932800/artifacts/build-resources/pods.json"
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply cmd/job-runtime-analyzer/main.go:11]: Well its a CLI arg

**Pull request merged**